python-lib/text_extraction/__init__.py

```diff
@@ -54,4 +54,108 @@ def extract_text_content(file_bytes, extension, with_pandoc):
+
+        try:
+            temporary_job_folder = os.getcwd()
+            with tempfile.NamedTemporaryFile(dir=temporary_job_folder, suffix=".{}".format(extension)) as tmp:
+                tmp.write(file_bytes)
+                markdown = pypandoc.convert_file(tmp.name, to="md", format=extension)
+        except Exception as e:
+            raise ValueError("Cannot convert file into markdown using pandoc because: {}".format(e))
+    
+        chunks = extract_markdown_chunks(markdown, filename)
+        return chunks
+
+
+def extract_markdown_chunks(markdown, filename):
+    # Code largely inspired by `langchain.text_splitter.HTMLHeaderTextSplitter`
+    lines = markdown.split("\n")
+    # Final output
+    lines_with_metadata = []
+    # text and metadata of the chunk currently being processed
+    current_text = []
+    current_metadata = {}
+    # Keep track of the nested header structure
+    # header_stack: List[Dict[str, Union[int, str]]] = []
+    header_stack = []
+    initial_metadata = {}
+
+    for line in lines:
+        stripped_line = line.strip()
+
+        # Check each line against each of the header types (e.g., #, ##), header_level is the number of '#'
+        for header_level in range(1, 4):
+            # Check if line starts with a header that we intend to split on
+            if stripped_line.startswith("#" * header_level) and (
+                # Header with no text OR header is followed by space are valid conditions that sep is being used a header
+                len(stripped_line) == header_level or stripped_line[header_level] == " "
+            ):
+
+                # Pop out headers of lower or same level from the stack
+                while header_stack and header_stack[-1]["level"] >= header_level:
+                    # We have encountered a new header at the same or higher level
+                    popped_header = header_stack.pop()
+                    # Clear the metadata for the popped header in initial_metadata
+                    if popped_header["level"] in initial_metadata:
+                        initial_metadata.pop(popped_header["level"])
+
+                # Push the current header to the stack
+                header = {
+                    "level": header_level,
+                    "data": stripped_line[header_level :].strip(),
+                }
+                header_stack.append(header)
+                # Update initial_metadata with the current header
+                initial_metadata[header_level] = header["data"]
+
+                # Add the previous line to the lines_with_metadata only if current_text is not empty
+                if current_text:
+                    lines_with_metadata.append(
+                        {
+                            "text": "\n".join(current_text),
+                            "metadata": current_metadata.copy(),
+                        }
+                    )
+                    current_text.clear()
+
+                break
+        else:
```

Comment:
Add a comment here to explain what the `for/else` syntax is used for. 
IIUC it's the case if we did not find any header in the current line ?